File: backend/src/main/java/org/akashihi/mdg/dao/BudgetRepository.java
package org.akashihi.mdg.dao;

import org.akashihi.mdg.entity.Budget;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.time.LocalDate;
import java.util.Optional;

@Repository
public interface BudgetRepository extends JpaRepository<Budget, Long> {
    Boolean existsByEndGreaterThanEqualAndBeginningLessThanEqual(LocalDate other_beginning, LocalDate other_end);
    Optional<Budget> findFirstByIdLessThanEqualOrderByIdDesc(Long id);
}

File: backend/src/main/kotlin/org/akashihi/mdg/dao/BudgetRepository.kt
package org.akashihi.mdg.dao

import org.akashihi.mdg.entity.Budget
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.JpaSpecificationExecutor
import org.springframework.stereotype.Repository
import java.time.LocalDate
import java.util.*

@Repository
interface BudgetRepository : JpaRepository<Budget, Long>, JpaSpecificationExecutor<Budget> {
    fun existsByEndGreaterThanEqualAndBeginningLessThanEqual(otherBeginning: LocalDate, otherEnd: LocalDate): Boolean
    fun existsByEndGreaterThanEqualAndBeginningLessThanEqualAndIdNot(otherBeginning: LocalDate, otherEnd: LocalDate, id: Long): Boolean
    fun findFirstByBeginningLessThanEqualAndEndGreaterThanEqual(beginning: LocalDate, end: LocalDate): Budget?
    fun findFirstByBeginningLessThanEqualOrderByBeginningDesc(beginning: LocalDate): Budget?
    fun findByEndGreaterThanEqualAndBeginningLessThanEqualOrderByBeginningAsc(from: LocalDate, to: LocalDate): Collection<Budget>
}
